pkg/redis/redisstore.go

```diff
@@ -0,0 +1,89 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+	"github.com/storacha-network/indexing-service/pkg/types"
+)
+
+// DefaultExpire is the expire time we set on Redis when Set/SetExpiration are called with expire=true
+const DefaultExpire = time.Hour
+
+// Client is a subset of functions from the golang redis client that we need to implement our cache
+type Client interface {
+	Get(context.Context, string) *redis.StringCmd
+	Set(ctx context.Context, key string, value interface{}, expiration time.Duration) *redis.StatusCmd
+	Expire(ctx context.Context, key string, expiration time.Duration) *redis.BoolCmd
+	Persist(ctx context.Context, key string) *redis.BoolCmd
+}
+
+// Store wraps the go redis client to implement our general purpose cache interface,
+// using the providedserialization/deserialization functions
+type Store[Key, Value any] struct {
+	fromRedis func(string) (Value, error)
+	toRedis   func(Value) (string, error)
+	keyString func(Key) string
+	client    Client
+}
+
+var (
+	_ Client                = (*redis.Client)(nil)
+	_ types.Cache[any, any] = (*Store[any, any])(nil)
+)
+
+// NewStore returns a new instance of a redis store with the provided serialization/deserialization functions
+func NewStore[Key, Value any](
+	fromRedis func(string) (Value, error),
+	toRedis func(Value) (string, error),
+	keyString func(Key) string,
+	client Client) *Store[Key, Value] {
+	return &Store[Key, Value]{fromRedis, toRedis, keyString, client}
+}
+
+// Get returns deserialized values from redis
+func (rs *Store[Key, Value]) Get(ctx context.Context, key Key) (Value, error) {
+	data, err := rs.client.Get(ctx, rs.keyString(key)).Result()
+
+	if err != nil {
+		var v Value
+		if err == redis.Nil {
+			return v, types.ErrKeyNotFound
+		}
+		return v, fmt.Errorf("error accessing redis: %w", err)
+	}
+	return rs.fromRedis(data)
+}
+
+// Set saves a serialized value to redis
+func (rs *Store[Key, Value]) Set(ctx context.Context, key Key, value Value, expires bool) error {
```

Comment:
What's the reason to use a boolean vs a duration for expiration?

Reply:
I dunno simplicity given it's an abstract interface. also the key distinction I'm thinking of is keys you can collect to keep the cache size down vs keys you dont want to do that on.